src/viam/app/data/client.py

```diff
@@ -290,10 +320,41 @@ async def bounding_box_labels_by_filter(self, filter: Optional[Filter] = None) -
         response: BoundingBoxLabelsByFilterResponse = await self._data_client.BoundingBoxLabelsByFilter(request, metadata=self._metadata)
         return response.labels
 
-    # TODO(RSDK-3637): Implement
-    async def data_capture_upload(self, metadata: UploadMetadata, sensor_contents: Optional[List[SensorData]] = None) -> None:
-        raise NotImplementedError()
+    async def data_capture_upload(self, metadata: UploadMetadata , sensor_contents: List[SensorData]) -> None:
+        """Upload binary or tabular sensor data.
+
+        Sync binary or tabular data collected on a robot through a specific component (i.e. a motor) along with the relevant metadata with
+        app.viam.com.
+
+        Args:
+            metadata (viam.app.proto.datasync.UploadMetaData): Metadata relating the data collected with the corresponding robot part. Must
+                specify a part ID, component type, component name, method name, and data type.
+            sensor_contents (List[viam.app.proto.datasync.SensorData]): A list of the data to be uploaded. If the metadata specifies a
```

Comment:
This is kind of finicky ("if _this_ field is set to _that_ value, then we upload `foo`, otherwise ..."). Ideally users shouldn't have to parse logic like that in docstrings. Probably would be better to just ask for people to pass the individual parts of `SensorData` (timestamps, and then data which is either `bytes` or `Mapping[str, Any]`) and then construct a proper `SensorData` object from that.

Reply:
I like this too I'll get it added.

Reply:
Do you think it might be awkward to ask for three lists that have to line up by index? the sensor_contents parameter is weird. If the user is uploading binary data, only a single SensorData object is required and allowed (i.e. one time_requested timestamp, one time_received timestamp, and then the binary data). On the other hand, if tabular binary data is being uploaded a list of SensorData files of any length > 0 is allowed, meaning we need to take in a list of time_requested timestamps and time_received timestamps and then the tabular data itself. Another caveat is that if it is tabular data, it has to ALL be tabular data. Do you think all of this is worth explaining in the docstrings? Or should we try something else. If we did this, I'm thinking the individual fields of the SensorData objects as parameters would look something like this:

foo(
    time_received: Union[TimeStamp, List[TimeStamp]],
    time_requested: Union[TimeStamp, List[TimeStamp]],
    data: Union[bytes, List[Mapping[str, Any]]]
).

Another thing to consider is that we still probably could avoid getting to verbose in the docstrings because there is pretty good error handling on the receiving end in app. The error messages make it pretty obvious if someone passes lots of binary data when they weren't supposed to, etc.

Reply:
Maybe we could break this down into two different methods, one for binary data and one for tabular data or something

Reply:
^ just to make it more clear when a list of timestamps is needed vs when just 1 is needed without having a ton of logic explained in the docstrings or even in the code